internal/util/ssa/managedfields.go

```diff
@@ -26,8 +26,75 @@ import (
 	"k8s.io/klog/v2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
 )
 
+const classicManager = "manager"
+
+// DropManagedFields modifies the managedFields entries on the object that belong to "manager" (Operation=Update)
+// to drop ownership of the given paths if there is no field yet that is managed by `ssaManager`.
+//
+// If we want to be able to drop fields that were previously owned by the "manager" we have to ensure that
+// fields are not co-owned by "manager" and `ssaManager`. Otherwise, when we drop the fields with SSA
+// (i.e. `ssaManager`) the fields would remain as they are still owned by "manager".
+// The following code will do a one-time update on the managed fields.
+// We won't do this on subsequent reconciles. This case will be identified by checking if `ssaManager` owns any fields.
+// Dropping ownership in paths for existing "manager" entries (which could also be from other controllers) is safe,
+// as we assume that if other controllers are still writing fields on the object they will just do it again and thus
+// gain ownership again.
+func DropManagedFields(ctx context.Context, c client.Client, obj client.Object, ssaManager string, paths []contract.Path) error {
+	// Return if `ssaManager` already owns any fields.
+	if hasFieldsManagedBy(obj, ssaManager) {
```

Comment:
q: is this function returning always true in the case of infra and boostrap machine?
can we make this smart and consider only paths? (this can be also a follow up)

Reply:
> q: is this function returning always true in the case of infra and boostrap machine?

No.

The following happens: (example based on KubeadmConfig)
1. After create:
    * manager owns all fields including labels and annotations
2. After DropManagedFields
    * manager owns all fields **except** labels and annotations
3. After updateExternalObject:
    * manager owns all fields **except** labels and annotations
    * capi-machineset owns labels and annotations

In 2 `hasFieldsManagedBy` returned false, after 3 it returns true.

So basically this func detects correctly if we have to cleanup managed fields again by checking if capi-machineset already wrote the object. Which is exactly what we want as far as I can tell.

In which way do you want to make it smarter? That this func detects if ssaManager already owns any/all passed in paths? Checking for all could be a problem in the general case when ssaManager does not write **all** paths that have been passed in.

It's fine in the MachineSet case though as we are even getting ownership of annotations if we don't have an opinion on a specific annotation:
```yaml
          f:annotations: {}
          f:labels:
            f:cluster.x-k8s.io/cluster-name: {}
            f:cluster.x-k8s.io/deployment-name: {}
            f:cluster.x-k8s.io/set-name: {}
            f:machine-template-hash: {}
            f:topology.cluster.x-k8s.io/deployment-name: {}
            f:topology.cluster.x-k8s.io/owned: {}
```

Reply:
thanks for the clarification